### ISSUE TITLE
Flag as suggestion when `await Task<T>` is silently converted to `async Task`

### ISSUE THREAD

**Issue by SteveDesmond-ca** [dotnet/roslyn-analyzers#5744]
### Describe the problem you are trying to solve

Consider the case:
```cs
public static async Task Main() => await Fail();
private static async Task<int> Fail() => await Task.FromResult(1);
```

While yes, `Task<T>` does inherit from `Task`, which makes this valid (aka it will compile) code, it does not do what's expected as it will not actually return the exit code from `Fail()`.

### Describe suggestions on how to achieve the rule

If introducing this suggestion globally seems like too much, this may be a specific case for exit codes from `Program.Main()` as that was specifically what bit me earlier this week.

### Additional context

dotnet/roslyn#57881 was originally filed in this repo, including this suggestion, but was sent over to dotnet/roslyn, and seems like it will be rejected there. It was suggested that this part of that issue be re-reported here again.

**Comment by svick** [comment]
I think this does exactly what it's supposed to and it's not even really about `async`, since the following code also works fine and does not warn:

```c#
public static void Main() => Fail();
private static int Fail() => 1;
```

So what you're saying is you want a warning when a return value from a function is ignored. That is something some languages care about (like F#), but I don't think it would be a good idea to introduce it in C# at this point, since it would have a lot of false positives.

And I don't think `Main` is special here: its return type clearly tells you whether it returns a value or not. I don't think it's common to have a `void`/`Task` method and expect it to return an `int`.

**Comment by Evangelink** [comment]
I might be wrong about the intent there, but it somehow feels like an extent of [CA1806](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1806), where the problem is that the returned "object" is ignored.

**Comment by Youssef1313** [comment]
@Evangelink The thing is that CA1806 is special cased for specific scenarios. I don't see how the reported scenario can fit into a non-general scenario. If CA1806 is generalized for every unused return value from a method, there will be MANY false positives.

**Comment by SteveDesmond-ca** [comment]
I think this is also exacerbated by the ambiguity of the expression body.

Using @svick's example:
```cs
public static void Main() => Fail();
private static int Fail() => 1;
```

Regardless of what the compiler actually does, what is my *intention* for `Main()`?
```cs
public static void Main()
{
	Fail();
}
```
vs
```cs
public static void Main()
{
	return Fail();
	//obviously this doesn't compile, but that's the point: if this was my intention, but the compiler did the other thing, maybe we should have a suggestion saying "hey, this might not work as expected"
}
```
The expression body syntax kind of conflates the two, which is how I hit this in the first place. I'd go so far as to say this is more about expression body ambiguity than return types or async/await.

Even my exact problem, written without an expression body, gives me the exact error I would have expected:
```cs
public static async Task Main()
{
	return await Task.FromResult(1);
}
```
So why not have the computer consider that maybe that's what was intended by
```cs
public static async Task Main() => await Task.FromResult(1);
```
	
There have also been several mentions of concerns around false positives, but there are ways to minimize those, including minimizing scope to expression body only (as that seems to be the scope of the problem), making it a suggestion (vs a warning), and making it opt-in (vs on by default).

**Comment by Evangelink** [comment]
> @Evangelink The thing is that CA1806 is special cased for specific scenarios. I don't see how the reported scenario can fit into a non-general scenario. If CA1806 is generalized for every unused return value from a method, there will be MANY false positives.

Obviously I am not suggesting to have it generic, otherwise I would recommend to move to F# where it's by design. As you said, CA1806 is specific for a list of cases where the return type **should not** be ignored. It seems to me that, putting aside the fact the returned value is ignored because of a cast, this is just another case that could be added to the rule. 

I am also thinking about a more global scenario (could be another rule) in which we would like to detect cases where a generic type is hidden through a non-generic (`Task<T>` -> `Task`, `ValueTask<T>` -> `ValueTask`...) but I am wondering if it's not too wide and could cause many FPs. We may want to exclude cases where it's losing the generic because of an interface to reduce potential FPs.

**Comment by mavasani** [comment]
FYI: We do have [IDE0058](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0058): `Remove unnecessary expression value`, which is hidden diagnostic/refactoring by default.

```c#
void M()
{
  Compute(); // IDE0058: computed value is never used.
}

int Compute();
```

**Comment by jmarolf** [comment]
I would be ok with a new rule that is essentially a special-cased version of [ca1806](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1806) that is explicitly about awaited values being ignored. 

if you do this:
```csharp
async Task M()
{
  await ComputeAsync(); 
}

Task<int> ComputeAsync() => Task.FromResult(1);
```

You should get a warning. At the very least invoking the async state-machine and not observing the result is a very odd and potentially dangerous thing to do. Yes someone could be doing this because they are trying to cause some side-effect to occur but I think offering the codefix of:

```csharp
async Task M()
{
  _ = await ComputeAsync(); 
}

Task<int> ComputeAsync() => Task.FromResult(1);
```

Would make it clear to both the analyzer and any future developers reading this code that ignoring this result is intentional. As @Youssef1313 points out I do not think this sort of strategy can be applied universally to expressions because you have various fluent style apis and horrible tragedies like `StringBuilder`. But I think this would be valuable if constrained only to await expressions.

**Comment by Youssef1313** [comment]
@mavasani @jmarolf Is this ready to implement?
